evmrpc/block.go

```diff
@@ -0,0 +1,109 @@
+package evmrpc
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/sei-protocol/sei-chain/x/evm/keeper"
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+	rpcclient "github.com/tendermint/tendermint/rpc/client"
+	"github.com/tendermint/tendermint/rpc/coretypes"
+)
+
+type BlockAPI struct {
+	tmClient    rpcclient.Client
+	keeper      *keeper.Keeper
+	ctxProvider func() sdk.Context
+	txDecoder   sdk.TxDecoder
+}
+
+func NewBlockAPI(tmClient rpcclient.Client, k *keeper.Keeper, ctxProvider func() sdk.Context, txDecoder sdk.TxDecoder) *BlockAPI {
+	return &BlockAPI{tmClient: tmClient, keeper: k, ctxProvider: ctxProvider, txDecoder: txDecoder}
+}
+
+func (a *BlockAPI) GetBlockByHash(ctx context.Context, blockHash common.Hash, fullTx bool) (map[string]interface{}, error) {
+	block, err := a.tmClient.BlockByHash(ctx, blockHash[:])
+	if err != nil {
+		return nil, err
+	}
+	blockRes, err := a.tmClient.BlockResults(ctx, &block.Block.Height)
+	if err != nil {
+		return nil, err
+	}
+	return encodeTmBlock(a.ctxProvider(), block, blockRes, a.keeper, a.txDecoder, fullTx)
+}
+
+func encodeTmBlock(
+	ctx sdk.Context,
+	block *coretypes.ResultBlock,
+	blockRes *coretypes.ResultBlockResults,
+	k *keeper.Keeper,
+	txDecoder sdk.TxDecoder,
+	fullTx bool,
+) (map[string]interface{}, error) {
+	number := big.NewInt(block.Block.Height)
+	blockhash := common.HexToHash(string(block.BlockID.Hash))
+	lastHash := common.HexToHash(string(block.Block.LastBlockID.Hash))
+	appHash := common.HexToHash(string(block.Block.AppHash))
+	txHash := common.HexToHash(string(block.Block.DataHash))
+	resultHash := common.HexToHash(string(block.Block.LastResultsHash))
+	miner := common.HexToAddress(string(block.Block.ProposerAddress))
+	gasLimit, gasWanted := int64(0), int64(0)
+	transactions := []interface{}{}
+	for _, txRes := range blockRes.TxsResults {
+		gasLimit += txRes.GasWanted
+		gasWanted += txRes.GasUsed
+		decoded, err := txDecoder(txRes.Data)
+		if err != nil {
+			fmt.Println(err)
+			return nil, errors.New("failed to decode transaction")
+		}
+		if len(decoded.GetMsgs()) != 1 {
+			// EVM message must have exactly one message
+			continue
```

Comment:
why not panic? or do we assume it's possible that we don't get strictly EVM msgs? if the latter, this seems quite brittle - is there a more robust way to detect for EVM msgs?

Reply:
the block may contain non-EVM transactions which would have more than 1 message. There are ante handlers checking for this so we won't have a block containing txs that have more than 1 message while one of the messages is EVM message